[#396] Add pgmoneta-walinfo CLI to read WAL files

sh-soltanpour opened the pull request:
### Add pgmoneta-walinfo CLI to read and display WAL files

This PR introduces the `pgmoneta-walinfo` command-line tool for reading and displaying information from PostgreSQL WAL files. Key features include:

- WAL file decompression
- JSON and raw output support
- Display with color palette
- Documentation updates for the new CLI

jesperpedersen commented:
Remember to squash

sh-soltanpour commented:
Hey @jesperpedersen, with some of the circular dependencies that I faced, I've decided to add sth like the `wal_abstractions.h` file that just contains general structs and flags used by different resource managers, wal reader, Walfile, and serializer. The file does not contain any method definitions and is just a place to define structs and flags used by different functionalities, structs such as `xlog_record`, `decoded_xlog_record` or even `walfile` used by wal_reader, serialzier and walfile. 

What do you think about this approach to avoid circular dependencies and be more future-proof?


![pgmoneta drawio](https://github.com/user-attachments/assets/24d5ae25-61bf-443e-a5ee-328adb1dbe54)

jesperpedersen commented:
Yeah, but could be in `walfile.h` ? Ideally it would be one `.h` that people have to learn.

`serialize.h` do we need that ? We have `read()` and `write()` and later we can change `struct walfile` through APIs...

Initial thoughts, I'll look at the code

Jubilee101 commented on src/libpgmoneta/walfile.c:
```diff
@@ -381,4 +383,39 @@ pgmoneta_destroy_walfile(struct walfile* wf)
+   {
+      struct json* record_json = NULL;
+      struct decoded_xlog_record* record = (struct decoded_xlog_record*) record_iterator->value->data;
+      pgmoneta_wal_display_decoded_record(&record_json, record, &config->servers[0]);
```
If you want to do one line per record here, just print out each of them in FORMAT_JSON_COMPACT here soon as you generate them. No need to put them in one big json and print out all together.

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,141 @@
+static void
+usage(void)
+{
+   printf("pgmoneta %s\n", VERSION);
```
pgmoneta-walinfo

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,141 @@
+   printf("\n");
+   printf("Options:\n");
+   printf("  -?, --help               Display help\n");
+   printf("  -v, --server-version     Specify PostgreSQL server version (13 to 17)\n");
```
Do we need this ?

sh-soltanpour replied:
hmm we can obtain it from the long_header I believe, will remove it

sh-soltanpour commented:
> Yeah, but could be in `walfile.h` ? Ideally it would be one `.h` that people have to learn.
> 
> `serialize.h` do we need that ? We have `read()` and `write()` and later we can change `struct walfile` through APIs...
> 
> Initial thoughts, I'll look at the code

Yeah I can put them all in `walfile.h` as well.

sh-soltanpour commented:
> `serialize.h` do we need that ? We have `read()` and `write()` and later we can change `struct walfile` through APIs...

For seralize, my idea is to take out the serialization logic from the `walfile.c` and assign it to another module(s), so there we can implement JSON, CSV, or anything else that may be needed in the future.

jesperpedersen commented:
We just have to figure out the best way to do it, since what we need are

```
WAL file -> CSV -> stdout
WAL file -> JSON -> stdout
```

@Jubilee101 is best to decide that...

Jubilee101 commented:
> For seralize, my idea is to take out the serialization logic from the walfile.c and assign it to another module(s), so there we can implement JSON, CSV, or anything else that may be needed in the future.

I do like the idea but for now maybe it's a bit overkill to have a separate module for that. And you don't really have to have multiple functions to convert it multiple types of instances. Just one `wal_record_serialize(walrecord* record, value_type type, value** value)` will probably do, type can be `ValueJson` for now, or `ValueCSV` which we don't have for now but can be added if needed. Then each type of values naturally comes with `to_string` for output.

That's my idea for now. I'll think about it.

sh-soltanpour commented:
> > For seralize, my idea is to take out the serialization logic from the walfile.c and assign it to another module(s), so there we can implement JSON, CSV, or anything else that may be needed in the future.
> 
> I do like the idea but for now maybe it's a bit overkill to have a separate module for that. And you don't really have to have multiple functions to convert it multiple types of instances. Just one `wal_record_serialize(walrecord* record, value_type type, value** value)` will probably do, type can just be `ValueJson` for now, or `ValueCSV` which we don't have currently but can be added later if needed. Then each type of values naturally comes with `to_string` for output.
> 
> That's my idea for now. I'll think about it.

Sounds good, thanks!
So I've implemented raw and JSON formats until now, and based on this I think we can implement CSV in another PR after ValueCSV is implemented.

I've also added the coloring for the raw printing format, but still need to adjust the code and colors and do a bit of cleanup on code and memory leakage.

jesperpedersen commented:
Do we need all of that in `walfile.h` ? It should only contain what is needed for higher levels, so `pgmoneta_describe_walfile` seem the only real change

jesperpedersen commented on src/libpgmoneta/walfile.c:
```diff
@@ -381,4 +379,120 @@ pgmoneta_destroy_walfile(struct walfile* wf)
+}
+
+void
+pgmoneta_wal_record_serialize(struct decoded_xlog_record* record, uint8_t magic_value, enum value_type type, struct value** value)
```
wal_reader ?

jesperpedersen commented on src/libpgmoneta/walfile.c:
```diff
@@ -381,4 +379,120 @@ pgmoneta_destroy_walfile(struct walfile* wf)
+}
+
+void
+pgmoneta_wal_record_display(struct decoded_xlog_record* record, uint8_t magic_value, enum value_type type)
```
wal_reader

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -83,10 +70,39 @@ is_bimg_apply(uint8_t bimg_info)
 
+int
+magic_value_to_postgres_version(uint16_t magic_value)
+{
```
magic value is specific for a release, so we don't need a range

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,160 @@
+   printf("\n");
+   printf("Options:\n");
+   printf("  -?, --help               Display help\n");
+   printf("  -v, --server-version     Specify PostgreSQL server version (13 to 17)\n");
```
`-v` is still there

jesperpedersen commented:
```
./pgmoneta-walinfo ~/PostgreSQL/pgmoneta/test/primary/wal/000000010000000000000002.zstd 
Missing --format argument
```

`raw` is default

jesperpedersen commented:
```
./pgmoneta-walinfo -F raw ~/PostgreSQL/pgmoneta/test/primary/wal/000000010000000000000002.zstd 
pgmoneta-walinfo: src/libpgmoneta/walfile/wal_reader.c:198: int pgmoneta_wal_parse_wal_file(char *, int, struct walfile *): Assertion `magic_value_to_postgres_version(long_header->std.xlp_magic) != -1' failed.
```

So, you need to support the compression algorithms

jesperpedersen commented:
```
rm: Standby, Rec/tot len: 50/50, Xid: 0, Info: 16, PrevLSN: 33554776 -- next_xid 752 latest_completed_xid 751 oldest_running_xid 752 -- (null)
```

->

```
Standby,50,50,0,16, 33554776,next_xid 752 latest_completed_xid 751 oldest_running_xid 752 -- (null)
```

What is `(null)` ?

Each column need a detailed description in the documentation in a walinfo section/chapter

jesperpedersen commented on src/include/utils.h:
```diff
@@ -46,6 +46,29 @@ extern "C" {
 #define LONG_TIME_LENGHT  16 + 1
 #define UTC_TIME_LENGTH   29 + 1
 
+// Define Windows 20 palette colors as constants using ANSI codes
```
Use `COLOR_`

jesperpedersen replied:
@Jubilee101 Do we want a separate `color.h` ?

sh-soltanpour commented:
> ./pgmoneta-walinfo -F raw ~/PostgreSQL/pgmoneta/test/primary/wal/000000010000000000000002.zstd

`(null)` is when there's no backup string information for that record, can replace it with an empty string. 
Sounds good, will take care of the rest as well

jesperpedersen commented on src/include/walfile.h:
```diff
@@ -30,9 +30,224 @@
 #include <deque.h>
-#include <walfile/wal_reader.h>
 
+#include <stdbool.h>
```
You need to describe why we need all of this in `walfile.h`...

None of it are used in the functions...

sh-soltanpour replied:
Yeah I remember my ide was complaining about not `bool` and some `int` not being defined so I added these, but now with the file changes it seems to be okay so I removed them.

jesperpedersen replied:
That is ok, but I meant all the changes in `walfile.h` - we only need the display function for now.

Remember that `walfile.h` should be as small as possible

sh-soltanpour replied:
I see. The reason for needing almost all of them is because we have this function`char* pgmoneta_wal_encode_xlog_record` in `walfile.c` used by writing functionality. 
But now I'm thinking about trying and bringing this function to `wal_reader` to reduce complexity in `walfile`.

sh-soltanpour replied:
ok I've been able to revert almost all of additions to `walfile.h` to make it as simple as possible, now it only has:

```
struct xlog_long_page_header_data;
struct walfile;
int pgmoneta_read_walfile(int server, char* path, struct walfile** wf);
int pgmoneta_write_walfile(struct walfile* wf, int server, char* path);
void pgmoneta_destroy_walfile(struct walfile* wf);
int pgmoneta_describe_walfile(char* path, enum value_type type);
```
I haven't pushed this yet

jesperpedersen replied:
That sounds great !

jesperpedersen replied:
Something like resource manager and record type will come into walfile.h later (2 steps from now - well, 3 including this one)

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,155 @@
+   printf("\n");
+
+   printf("Usage:\n");
+   printf("  ./pgmoneta-walinfo [FILE_PATH] --format [raw|json]\n");
```
You just need `pgmoneta-walinfo <file>` here

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,155 @@
+usage(void)
+{
+   printf("pgmoneta-walinfo %s\n", VERSION);
+   printf("  Command line utility to read write-ahead log files\n");
```
Do `Write-Ahead Log (WAL) files`

sh-soltanpour commented:
> ```
> ./pgmoneta-walinfo -F raw ~/PostgreSQL/pgmoneta/test/primary/wal/000000010000000000000002.zstd 
> pgmoneta-walinfo: src/libpgmoneta/walfile/wal_reader.c:198: int pgmoneta_wal_parse_wal_file(char *, int, struct walfile *): Assertion `magic_value_to_postgres_version(long_header->std.xlp_magic) != -1' failed.
> ```
> 
> So, you need to support the compression algorithms.
> 
> You can decompress into `/tmp/`

Should we assume the compression based on the file extension? or for example, ask the user to enter `--compressed`.
I'm thinking the first one is better but just wanted to confirm with you

jesperpedersen commented:
Yeah, just look at the extension name - that is what we do right now

jesperpedersen commented:
Well, we can decide what `raw` means once we have CSV... so just leave it for now

jesperpedersen commented:
Documentation for walinfo is really important - and what the columns and colors mean

jesperpedersen commented:
LSNs should be encoded as LSNs, so `X/Y` format

sh-soltanpour commented:
@jesperpedersen One thing I noticed is that decompression methods, such as `pgmoneta_zstandardd_file` are removing the source file after decompression.
In our case, I think this is an unwanted side effect and we do not necessarily want to remove them. I was wondering where the deletion functionality is used and maybe we can separate it from the decompression function and do it explicitly where needed.

jesperpedersen commented:
For now, just use `pgmoneta_copy_file(..., NULL)` to `/tmp/`,  display, then delete with `pgmoneta_delete_file(..., NULL)`

jesperpedersen commented:
See if you can do the format as

```
resource_manager |  start_lsn |  end_lsn  | xid | record_type  | record_length | description
```

-- https://www.percona.com/blog/see-whats-inside-a-wal-file-using-pg_waldump-and-pg_walinspect/

jesperpedersen commented:
You can likely do a `char* pgmoneta_lsn_to_string()` in `utils.h`

sh-soltanpour commented:
> |

I've changed it to be like this:

`Resource manager | rec len | tot len | xid | prev lsn | description (data and backup)`
Currently I'm not tracking the current LSN for each record to print, if that's okay, I can work on tracking it and adding it later.

jesperpedersen commented:
We can take it step-by-step.

Lets focus on compression support next

jesperpedersen commented:
```
2024-10-19 06:12:26 ERROR compresssion.c:70 pgmoneta_decompress: no decompression callback found for file /tmp/compressed_wal
2024-10-19 06:12:26 FATAL walfile.c:228 Failed to decompress WAL file at test/primary/wal/000000010000000000000002.zstd
Error while reading/describing WAL file
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1965077==ERROR: AddressSanitizer: SEGV on unknown address 0x55a8edc13930 (pc 0x55a8edb28be6 bp 0x000000000000 sp 0x7ffda201dfe0 T0)
==1965077==The signal is caused by a WRITE memory access.
    #0 0x55a8edb28be6 in __asan::Allocator::Deallocate(void*, unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) (build/src/pgmoneta-walinfo+0x2fbe6) (BuildId: 049ca8c3d2d4211e4938caa73b63f8ac3232df2f)
    #1 0x55a8edbbffff in free (official/build/src/pgmoneta-walinfo+0xc6fff) (BuildId: 049ca8c3d2d4211e4938caa73b63f8ac3232df2f)
    #2 0x55a8edbfef1a in main src/walinfo.c:138:7
```

jesperpedersen commented on src/cli.c:
```diff
@@ -1745,7 +1745,6 @@ translate_server_retention_argument(struct json* response, char* tag)
    }
 }
 
-
```
Just leave this such that the patch is smaller

jesperpedersen commented:
```
Standby | 50 | 50 | 0 | 0/2000158 | next_xid 752 latest_completed_xid 751 oldest_running_xid 752 
XLOG | 30 | 30 | 0 | 0/3000028 |  
Standby | 50 | 50 | 0 | 0/3000060 | next_xid 752 latest_completed_xid 751 oldest_running_xid 752 
XLOG | 114 | 114 | 0 | 0/3000080 | redo 0/3000060; tli 1; prev tli 1; fpw true; wal_level replica; xid 0:752; oid 16391; multi 1; offset 0; oldest xid 730 in DB 1; oldest multi 1 in DB 5; oldest/newest commit timestamp xid: 0/0; oldest running xid 752 online 
Standby | 50 | 50 | 0 | 0/30000B8 | next_xid 752 latest_completed_xid 751 oldest_running_xid 752 
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1965948==ERROR: AddressSanitizer: SEGV on unknown address 0x55f4b3460930 (pc 0x55f4b3375be6 bp 0x000000000000 sp 0x7ffd285a5400 T0)
==1965948==The signal is caused by a WRITE memory access.
    #0 0x55f4b3375be6 in __asan::Allocator::Deallocate(void*, unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) (build/src/pgmoneta-walinfo+0x2fbe6) (BuildId: 049ca8c3d2d4211e4938caa73b63f8ac3232df2f)
    #1 0x55f4b340cfff in free (build/src/pgmoneta-walinfo+0xc6fff) (BuildId: 049ca8c3d2d4211e4938caa73b63f8ac3232df2f)
    #2 0x55f4b344bf1a in main src/walinfo.c:138:7
```

jesperpedersen commented on src/include/utils.h:
```diff
@@ -46,6 +46,29 @@ extern "C" {
+#define COLOR_MONEY_GREEN "\033[32m"   /* Close approximation */
+#define COLOR_SKY_BLUE "\033[36m"      /* Close approximation */
+#define COLOR_CREAM "\033[97m"         /* Close approximation */
+#define COLOR_MEDIUM_GREY "\033[90m"
```
Align the values to this line

jesperpedersen commented on src/libpgmoneta/compresssion.c:
```diff
@@ -0,0 +1,74 @@
+   }
+   else
+   {
+      *decompress_cb = NULL;
```
Start with this.

Look at `int` as a return value in case `path` is `NULL` or unknown

jesperpedersen commented on src/libpgmoneta/compresssion.c:
```diff
@@ -0,0 +1,74 @@
+{
+   compression_func decompress_cb = NULL;
+   pgmoneta_decompression_file_callback(from, &decompress_cb);
+   if (decompress_cb == NULL)
```
Then you can clean this up

jesperpedersen commented on src/libpgmoneta/walfile.c:
```diff
@@ -381,4 +203,56 @@ pgmoneta_destroy_walfile(struct walfile* wf)
+   }
+
+   // Based on the file extension, if it's a compressed file, decompress it in /tmp
+   char* decompressed_wal_path = NULL;
```
Start the function with all the declarations

jesperpedersen commented on src/libpgmoneta/walfile.c:
```diff
@@ -381,4 +203,56 @@ pgmoneta_destroy_walfile(struct walfile* wf)
+   }
+   pgmoneta_deque_iterator_destroy(record_iterator);
+   pgmoneta_destroy_walfile(wf);
+   return 0;
```
Do the standard

```
   free(...);

   return 0;

error:

   free(...);

   return 1;
}

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -186,6 +195,18 @@ pgmoneta_wal_parse_wal_file(char* path, struct server* server_info, struct walfi
    wal_file->long_phd = long_header;
+   assert(magic_value_to_postgres_version(long_header->std.xlp_magic) != -1);
+
+   if (server == -1)
```
Do we need this ?

Getting the magic number from the file should be enough...

sh-soltanpour replied:
I need it because I want to add server info to the shared memory, it's also used outside of the scope of this function, and I need to know whether to write it in the index 0 or another index if it's already defined.

jesperpedersen replied:
That is ok - lets assume that the WAL file "fits"

sh-soltanpour replied:
> That is ok - lets assume that the WAL file "fits"

Hmm sorry not sure if I quite understood what you mean.
In general, I need the sever info outside of the scope of this function and I think it's easier to free it when with the shared memory at the end.

jesperpedersen replied:
I mean that the WAL file belongs to the server in question such that the OIDs, relfillnodes, ... match up

sh-soltanpour replied:
ok so anything needed to do here?

jesperpedersen replied:
we can always circle back to it

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -817,3 +814,270 @@ pgmoneta_wal_get_record_block_tag_extended(struct decoded_xlog_record* pRecord,
+      free(rm_desc);
+      free(backup_str);
+      free(prev_lsn_string);
+   }
```
It is a lot easier to read if you do the declarations first, and always end with `free(...); return;`

We can discuss the `ValueString` formatting later... f.ex. it is important that `RmgrTable[record->header.xl_rmid].name` has its own color

sh-soltanpour replied:
I've put the definitions at the beginning. 
since there are two different paths and to avoid having a lot of
`if (x == NULL) free(x);` at the end, I've put the freeing logic at the end of each branch

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,150 @@
+   printf("\n");
+
+   printf("Usage:\n");
+   printf("  ./pgmoneta-walinfo [FILE_PATH]\n");
```
`./`: Remove
`[FILE_PATH]` -> `<file>`

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,150 @@
+   printf("  ./pgmoneta-walinfo [FILE_PATH]\n");
+   printf("\n");
+   printf("Options:\n");
+   printf("  -F, --format             Output format (raw, json)\n");
```
There only need to be one space after the longest help line to the description, and then align all help lines to that

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,150 @@
+   }
+
+   return 0;
+}
```
Standard pattern w/ `free(); return 0; error: free(); return 1;`

sh-soltanpour commented:
> ```
> Standby | 50 | 50 | 0 | 0/2000158 | next_xid 752 latest_completed_xid 751 oldest_running_xid 752 
> XLOG | 30 | 30 | 0 | 0/3000028 |  
> Standby | 50 | 50 | 0 | 0/3000060 | next_xid 752 latest_completed_xid 751 oldest_running_xid 752 
> XLOG | 114 | 114 | 0 | 0/3000080 | redo 0/3000060; tli 1; prev tli 1; fpw true; wal_level replica; xid 0:752; oid 16391; multi 1; offset 0; oldest xid 730 in DB 1; oldest multi 1 in DB 5; oldest/newest commit timestamp xid: 0/0; oldest running xid 752 online 
> Standby | 50 | 50 | 0 | 0/30000B8 | next_xid 752 latest_completed_xid 751 oldest_running_xid 752 
> AddressSanitizer:DEADLYSIGNAL
> =================================================================
> ==1965948==ERROR: AddressSanitizer: SEGV on unknown address 0x55f4b3460930 (pc 0x55f4b3375be6 bp 0x000000000000 sp 0x7ffd285a5400 T0)
> ==1965948==The signal is caused by a WRITE memory access.
>     #0 0x55f4b3375be6 in __asan::Allocator::Deallocate(void*, unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) (build/src/pgmoneta-walinfo+0x2fbe6) (BuildId: 049ca8c3d2d4211e4938caa73b63f8ac3232df2f)
>     #1 0x55f4b340cfff in free (build/src/pgmoneta-walinfo+0xc6fff) (BuildId: 049ca8c3d2d4211e4938caa73b63f8ac3232df2f)
>     #2 0x55f4b344bf1a in main src/walinfo.c:138:7
> ```

These must be fixed now as well

sh-soltanpour commented:
@jesperpedersen for documentations, I guess I need to:
1- add man to `doc/man`
2- add explanations to `dev-06-wal.md`
3- `Developers.md`

Is there anywhere else that I should update as well?

jesperpedersen commented:
And, `WAL.md` and put it into the `pgmoneta.spec` file

jesperpedersen commented:
Rebase, please

jesperpedersen commented on doc/DEVELOPERS.md:
````diff
@@ -362,6 +362,21 @@ pgmoneta-cli -c pgmoneta.conf status details
+```sh
+pgmoneta-walinfo [ -F FORMAT ] <file>
+```
+- `-F, --format`: Specify the output format (`raw` or `json`).
````
Just the whole thing in one comment block

jesperpedersen commented on doc/manual/dev-06-wal.md:
````diff
@@ -6,6 +6,53 @@
+pgmoneta-walinfo [ -F FORMAT ] <file>
+```
+
+- `-F, --format`: Specify the output format (`raw` or `json`). Default is `raw`.
````
Same here

jesperpedersen commented on src/walinfo.c:
```diff
@@ -0,0 +1,162 @@
+   }
+   else
+   {
+      fprintf(stderr, "Missing FILE_PATH argument\n");
```
`[FILE_PATH]` -> `<file>`

sh-soltanpour commented:
@jesperpedersen  I've fixed the issue about the record being split across two files. With the file you sent, I've also found another bug in `rm_database` that was updated in v17 so I patched that one as well.

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -60,18 +60,6 @@ print_long_page_header(struct xlog_long_page_header_data* header)
    pgmoneta_log_trace("-----------------\n");
```
You don't need `\n` for `pgmoneta_log_XYZ` functions

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -130,27 +138,28 @@ read_all_page_headers(FILE* file, struct xlog_long_page_header_data* long_header
+finish:
    return;
 error:
    pgmoneta_log_fatal("Error: Could not read all page headers\n");
```
Here is one

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -279,14 +308,17 @@ pgmoneta_wal_parse_wal_file(char* path, struct server* server_info, struct walfi
-      if (pgmoneta_wal_decode_xlog_record(buffer, decoded, record, long_header->xlp_xlog_blcksz, server_info))
+      if (pgmoneta_wal_decode_xlog_record(buffer, decoded, record, long_header->xlp_xlog_blcksz, long_header->std.xlp_magic))
       {
          pgmoneta_log_fatal("error in decoding\n");
```
Here is one

jesperpedersen replied:
They are in src/libpgmoneta/walfile.c and src/libpgmoneta/walfile/wal_reader.c

sh-soltanpour replied:
sorry sorry I don't understand this and the one above, can you please elaborate?

sh-soltanpour replied:
ah I see, you mean \n in the pgmoneta_logs, fixed!

jesperpedersen commented on src/libpgmoneta/walfile/wal_reader.c:
```diff
@@ -279,14 +308,17 @@ pgmoneta_wal_parse_wal_file(char* path, struct server* server_info, struct walfi
-         pgmoneta_wal_display_decoded_record(decoded, ++count, server_info);
+         if (decoded == NULL)
+         {
+            printf("decoded is NULL\n");
```
-> `pgmoneta_log_fatal`

jesperpedersen commented:
Remember to rebase

jesperpedersen commented:
```
Heap | 80 | 80 | 754 | 0/3FFFF40 | off 130 flags 0x00 blkref #0: rel 1663/16391/16393 forknum 0 blk 1116
Incomplete | Skipped
```

We need to have the same number of `|` for a CSV format.

Colors are still all red until description

jesperpedersen commented:
```
Heap | 80 | 80 | 754 | 0/4FFFF40 | off 61 flags 0x00 blkref #0: rel 1663/16391/16393 forknum 0 blk 2654
Incomplete | Skipped
AddressSanitizer:DEADLYSIGNAL
=================================================================
==618467==ERROR: AddressSanitizer: SEGV on unknown address (pc 0x5566a143cbf6 bp 0x000000000000 sp 0x7ffcdb7a9da0 T0)
==618467==The signal is caused by a READ memory access.
==618467==Hint: this fault was caused by a dereference of a high value address (see register values below).  Disassemble the provided pc to learn which register was used.
    #0 0x5566a143cbf6 in __asan::Allocator::Deallocate(void*, unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType) (official/build/src/pgmoneta-walinfo+0x2fbf6) (BuildId: 206b60b9df1c891aeef27c5fab45ec15bc9b284c)
    #1 0x5566a14d400f in free (build/src/pgmoneta-walinfo+0xc700f) (BuildId: 206b60b9df1c891aeef27c5fab45ec15bc9b284c)
    #2 0x7ff8879a0e1a in pgmoneta_destroy_walfile src/libpgmoneta/walfile.c:177:10
    #3 0x7ff8879a1981 in pgmoneta_describe_walfile src/libpgmoneta/walfile.c:260:4
    #4 0x5566a1512e3b in main src/walinfo.c:135:11
```

sh-soltanpour commented:
> ```
> Heap | 80 | 80 | 754 | 0/3FFFF40 | off 130 flags 0x00 blkref #0: rel 1663/16391/16393 forknum 0 blk 1116
> Incomplete | Skipped
> ```
> 
> We need to have the same number of `|` for a CSV format.
> 
> Colors are still all red until description

Ok sounds good, also working on fixing the JSON output to be a single JSON.

jesperpedersen commented:
If it is easier to pick a color for each column right now, then just do that

sh-soltanpour commented:
@jesperpedersen, @Jubilee101  If I want to preserve both your comments about the final json file being a valid JSON and also not to create a huge JSON in memory and print it, the code would be like this:

```
if (type == ValueJSON)
   {
       printf("{ \"WAL\": [\n");
       int count = 0;
       while (pgmoneta_deque_iterator_next(record_iterator))
       {
           printf("{\"Record\": ");
           record = (struct decoded_xlog_record*) record_iterator->value->data;
           pgmoneta_wal_record_display(record, wf->long_phd->std.xlp_magic, type);
           printf("}");
           if (++count < pgmoneta_deque_size(wf->records))
           {
               printf(",\n");
           }
       }
       printf("]\n}");
   }
   else
   {
       while (pgmoneta_deque_iterator_next(record_iterator))
       {
           record = (struct decoded_xlog_record*) record_iterator->value->data;
           pgmoneta_wal_record_display(record, wf->long_phd->std.xlp_magic, type);
       }
   }
```

jesperpedersen commented:
Lets not worry about the memory for now.

Create a structure such that you can just call 

```
char*
pgmoneta_json_to_string(struct json* object, int32_t format, char* tag, int indent);
```

aka `pgmoneta_json_to_string(wal_son, FORMAT_JSON, NULL, 0);`

jesperpedersen commented:
Any valid JSON is ok now though, so you can use your method as well

jesperpedersen commented:
Use your method; we have bigger fish to fry... we can circle back

sh-soltanpour commented:
Ok so I've done these:

1- JSON output. I think there's a problem with our JSON printer, it's truncating long strings, still need to check this. 
2- Fixed errors on WAL files Jesper sent: Needed to handle EOF in some other parts as well.
3- Colors: Added a color per column.
4- Put the same number of `|` for incomplete records.

jesperpedersen commented:
Merged.

Thanks for your contribution !
